### PR TITLE
Refactor reporting init claim

### PR DESCRIPTION
**Pull request by ochaloup**
Reporting and refactoring changes.

~~I created this as "DRAFT" but I consider it finished for review and as the last part of the follow-up tasks for the settlement pipeline creation. There could be more tuning but after aligning the PR with PR reviews I would consider the follow-up tasks as ready for the current time.~~

~~I create this as "DRAFT" because it depends on https://github.com/marinade-finance/validator-bonds/pull/54 When PR#54 is processed I would rebase this one and consider it as fully prepared for review.~~

### CODE REVIEW

File: settlement-pipelines/src/bin/claim_settlement.rs [hunk 1]
```diff
@@ -99,491 +118,370 @@ async fn main() -> anyhow::Result<()> {
...
             })?;
         insert_merkle_tree_loaded_data(loaded_data, None, Some(settlement_collection))?;
     }
 
     Ok(())
 }
 
+#[allow(clippy::too_many_arguments)]
+async fn claim_settlement<'a>(
+    program: &Program<Rc<DynSigner>>,
```
Comment (janlegner) [inline settlement-pipelines/src/bin/claim_settlement.rs:304]: RC again - can you please try to find all occurrences in the code?

Comment (ochaloup) [reply settlement-pipelines/src/bin/claim_settlement.rs:304]: I know about it and will do the fix after this get processed. I created a separate task to fix this issue.
My plan is to finish the pipelines refactor and reporting fixes (this PR), then to develop the bonds config update and then to fix the issue with usage of Rc builder and pipeline code.

**Comment by ochaloup** [comment]
@janlegner the code was aligned to prior PR fixes and tested in pipelines. It's ready to be reviewed (and then potentially pipeline running).
I'm aware of the issue with `Rc` that I will address in another PR.

**Comment by ochaloup** [comment]
Updated on rebase against `main`.

File: .buildkite/claim-settlements.yml [hunk 2]
```diff
@@ -60,14 +60,22 @@ steps:
       RUST_LOG: info,solana_transaction_builder_executor=debug,solana_transaction_builder=debug,builder_executor=debug,solana_transaction_executor=debug,settlement_pipelines=debug,claim_settlement=debug
       # RUST_BACKTRACE: full
     commands:
+    - '. "$HOME/.cargo/env"'
     - |
       claimable_epochs_num=$(buildkite-agent meta-data get claimable_epochs_num)
-      [ "$$claimable_epochs_num" -eq 0 ] && echo 'No settlement to claim from' && exit 0
+      if [ "$$claimable_epochs_num" -eq 0 ]; then
```
Comment (janlegner) [inline .buildkite/claim-settlements.yml:66]:
```suggestion
      if [[ $$claimable_epochs_num -eq 0 ]]; then
```

File: .buildkite/claim-settlements.yml [hunk 3]
```diff
@@ -76,26 +84,70 @@ steps:
...
+          notification_result="SUCCEEDED"
+          notification_color="52224"
+        fi
+     - 'buildkite-agent meta-data set notification_color "$$notification_color"'
+     - 'buildkite-agent meta-data set notification_result "$$notification_result"'
+     - 'echo "Notification result: $$notification_result, color: $$notification_color"'
+     - 'mkdir ./reports'
+     - 'buildkite-agent artifact download --include-retried-jobs "build-report.*" ./reports/'
+     - 'report_path="./reports/$(ls -v1 reports/ | tail -n 1)"'
+     - '[ -f "$$report_path" ] || echo "No report generated; UNKNOWN ERROR" >> "$$report_path"'
```
Comment (janlegner) [inline .buildkite/claim-settlements.yml:115]:
```suggestion
     - '[[ -f "$$report_path" ]] || echo "No report generated; UNKNOWN ERROR" >> "$$report_path"'
```

File: .buildkite/init-settlements.yml [hunk 4]
```diff
@@ -83,60 +83,58 @@ steps:
...
      - 'buildkite-agent meta-data set notification_result "$$notification_result"'
      - 'echo "Notification result: $$notification_result, color: $$notification_color"'
      - 'mkdir ./reports'
      - 'buildkite-agent artifact download --include-retried-jobs "build-report.*" ./reports/'
      - 'report_path="./reports/$(ls -v1 reports/ | tail -n 1)"'
-     - '[ -f "$$report_path" ] || echo "UNKNOWN ERROR" >> "$$report_path"'
-     - 'cp "$$report_path" ./init-settlement-report.txt'
-     - 'gcloud storage cp ./init-settlement-report.txt "$$gs_bucket/$$epoch/init-settlement-report.$(date +%s).txt"'
-     - 'buildkite-agent meta-data set attempts_count "$(grep -c "#ATTEMPT" "$$report_path")"'
+     - '[ -f "$$report_path" ] || echo "No report generated; UNKNOWN ERROR" >> "$$report_path"'
```
Comment (janlegner) [inline .buildkite/init-settlements.yml:131]:
```suggestion
     - '[[ -f "$$report_path" ]] || echo "No report generated; UNKNOWN ERROR" >> "$$report_path"'
```

File: settlement-pipelines/README.md [hunk 5]
```diff
@@ -3,8 +3,30 @@
 Set of CLI binaries that works as a pipeline off-chain
 management for the [Validator Bonds Program](../programs/validator-bonds/README.md).
 
+## Provided commands
+
+* [init-settlement](./src/bin/init_settlement.rs) : from provided JSON it creates the `Settlement` accounts on-chain
+* [list-claimable-epoch](./src/bin/list_claimable_epoch.rs) : printing a list of epochs that contains some `Settlement` that is possible to claim it
+* [claim-settlement](./src/bin/claim_settlement.rs) : searching on-chain for settlements to be claimed and claiming based on provided JSONs with merkle proofs
+* [list-settlement](./src/bin/list_settlement.rs) : from provided JSON files it derives `Settlement` accounts addresses and prints it out
+* [close-settlement](./src/bin/close_settlement.rs) : checking chain for `Settlement`s to be possible to close and resetting stake accounts, using provided list of Settlement addresses to search for the Settlement stake authorities
```
Comment (janlegner) [inline settlement-pipelines/README.md:12]:
```suggestion
## Provided commands

* [init-settlement](./src/bin/init_settlement.rs) : from provided JSON it creates the `Settlement` accounts on-chain
* [list-claimable-epoch](./src/bin/list_claimable_epoch.rs) : printing a list of epochs that contains some `Settlement` that is possible to claim it
* [claim-settlement](./src/bin/claim_settlement.rs) : searching on-chain for settlements to be claimed and claiming based on provided JSONs with merkle proofs
* [list-settlement](./src/bin/list_settlement.rs) : from provided JSON files it derives `Settlement` accounts addresses and prints it out
* [close-settlement](./src/bin/close_settlement.rs) : checking chain for `Settlement`s to be possible to close and reset stake accounts, using a provided list of Settlement addresses to search for the Settlement stake authorities
```

File: settlement-pipelines/README.md [hunk 6]
```diff
@@ -3,8 +3,30 @@
...
+There are 3 pipelines in use for the binary commands.
+
+* [init-settlements](../.buildkite/init-settlements.yml) : initializing Settlements for an epoch based on generated JSON files.
+  The pipeline is expected to be called after the JSON files are generated at [prepare-claims](../.buildkite/prepare-claims.yml)
+* [claim-settlements](../.buildkite/claim-settlements.yml) : claiming settlements when possible. It's executed as a cron job
+  once a time. It checks the on-chain state if possible to claim some settlement. The Settlement can be claimed
+  in time range of < `Settlement Creation + non-claimable-slots` - `Settlement Creation Epoch - Config claimable epoch`>.
+* [close-settlements](../.buildkite/close-settlements.yml) : closing `Settlement` accounts, closing `SettlementClaim` accounts,
+  and resetting state of stake accounts to be associated back to validator `Bond` when not claimed.
+  It verifies the on-chain state if the Settlement expires and if not exists the `SettlementClaim` can be closed.
```
Comment (janlegner) [inline settlement-pipelines/README.md:25]:
```suggestion
## Pipeline usage

There are 3 pipelines in use for the binary commands.

* [init-settlements](../.buildkite/init-settlements.yml) : initializing Settlements for an epoch based on generated JSON files.
  The pipeline is expected to be called after the JSON files are generated at [prepare-claims](../.buildkite/prepare-claims.yml)
* [claim-settlements](../.buildkite/claim-settlements.yml) : claiming settlements when possible. It's executed as a cron job
  once a time. It checks the on-chain state if possible to claim some settlement. The Settlement can be claimed
  in the time range of < `Settlement Creation + non-claimable-slots` - `Settlement Creation Epoch - Config claimable epoch`>.
* [close-settlements](../.buildkite/close-settlements.yml) : closing `Settlement` accounts, closing `SettlementClaim` accounts,
  and resetting the state of stake accounts to be associated back to validator `Bond` when not claimed.
  It verifies the on-chain state if the Settlement expires and if it does not exist the `SettlementClaim` can be closed.
```

File: settlement-pipelines/src/reporting.rs [hunk 7]
```diff
@@ -20,11 +22,10 @@ impl<T: PrintReportable> ReportHandler<T> {
         }
     }
 
-    pub async fn report(&self) -> anyhow::Result<()> {
+    pub async fn print_report(&self) {
```
Comment (janlegner) [inline settlement-pipelines/src/reporting.rs:25]: what is the benefit of having this async?

File: settlement-pipelines/src/bin/claim_settlement.rs [hunk 8]
```diff
@@ -99,491 +118,370 @@ async fn main() -> anyhow::Result<()> {
...
+            .iter()
+            .map(|p| p.to_str())
+            .collect::<Vec<_>>(),
+        claiming_data.keys().collect::<Vec<_>>()
+    );
+    Ok(claiming_data)
 }
 
-fn process_merkle_trees_file(
+fn process_json_file(
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply settlement-pipelines/src/bin/claim_settlement.rs]: A bit of leftover from multiple refactorings. No reason to be that way. PR update. Thanks for pointing this out.

**Review by janlegner — APPROVED** [review]

**Pull request merged**